src/pepr/operator/controllers/keycloak/client-sync.ts

```diff
@@ -0,0 +1,187 @@
+import { K8s, Log, fetch, kind } from "pepr";
+
+import { UDSConfig } from "../../../config";
+import { Store } from "../../common";
+import { Sso, UDSPackage } from "../../crd";
+import { Client } from "./types";
+
+const apiURL =
+  "http://keycloak-http.keycloak.svc.cluster.local:8080/realms/uds/clients-registrations/default";
+
+/**
+ * Create or update the Keycloak clients for the package
+ *
+ * @param pkg the package to process
+ *
+ * @returns the list of client refs
+ */
+export async function keycloak(pkg: UDSPackage) {
+  // Get the list of clients from the package
+  const clientReqs = pkg.spec?.sso || [];
+  const refs: string[] = [];
+
+  // Pull the isAuthSvcClient prop as it's not part of the KC client spec
+  for (const clientReq of clientReqs) {
+    const ref = await syncClient(clientReq, pkg);
+    refs.push(ref);
+  }
+
+  await purgeSSOClients(pkg, refs);
+
+  return refs;
+}
+
+/**
+ * Remove any remaining clients that are not in the refs list
+ *
+ * @param pkg the package to process
+ * @param refs the list of client refs to keep
+ */
+export async function purgeSSOClients(pkg: UDSPackage, refs: string[] = []) {
+  // Check for any clients that are no longer in the package and remove them
+  const currentClients = pkg.status?.ssoClients || [];
+  const toRemove = currentClients.filter(client => !refs.includes(client));
+  for (const ref of toRemove) {
+    const token = Store.getItem(ref);
+    const clientId = ref.replace("sso-client-", "");
+    if (token) {
+      await apiCall({ clientId }, "DELETE", token);
+    } else {
+      Log.warn(pkg.metadata, `Failed to remove client ${clientId}, token not found`);
+    }
+  }
+}
+
+async function syncClient(
+  { isAuthSvcClient, ...clientReq }: Sso,
+  pkg: UDSPackage,
+  isRetry = false,
+) {
+  Log.debug(pkg.metadata, `Processing client request: ${clientReq.clientId}`);
+
+  try {
+    // Not including the CR data in the ref because Keycloak client IDs must be unique already
+    const name = getClientName(clientReq);
+    const token = Store.getItem(name);
+
+    let client: Client;
+
+    // If the redirectUris are not set, default to *
+    if (!clientReq.redirectUris) {
+      clientReq.redirectUris = ["*"];
```

Comment:
This is a pretty insecure default. Can we just make the field required?

Reply:
yep no prob